File: src/lib/game/imageService.ts
/**
 * ImageService
 * ------------
 * Clean image-selection service (mirrors backend/app/services/image_service.py).
 *
 * `get_random_image()` returns metadata for a configurable set of puzzle
 * images. The registry intentionally contains ONLY metadata – never any
 * solution / ordering information. During the PUZZLE phase players are not
 * given the full image URL; they receive individually cropped pieces from the
 * authenticated /piece endpoint instead.
 */
import type { ImageMeta } from "./types";

const supabaseUrl =
  process.env.NEXT_PUBLIC_SUPABASE_URL ??
  "https://dewdcxssvkvbgyyenmmu.supabase.co";

function getStorageOriginalUrl(slug: string): string {
  return `${supabaseUrl}/storage/v1/object/public/puzzle-images/${slug}/original.webp`;
}

const IMAGES: ImageMeta[] = [
  {
    id: "image_001",
    slug: "cosmic-fox",
    url: getStorageOriginalUrl("cosmic-fox"),
    name: "Cosmic Fox",
  },
  {
    id: "image_002",
    slug: "futuristic-city",
    url: getStorageOriginalUrl("futuristic-city"),
    name: "Futuristic City",
  },
  {
    id: "image_003",
    slug: "space-explorer",
    url: getStorageOriginalUrl("space-explorer"),
    name: "Space Explorer",
  },
  {
    id: "image_004",
    slug: "neon-cyberpunk",
    url: getStorageOriginalUrl("neon-cyberpunk"),
    name: "Neon Cyberpunk",
  },
  {
    id: "image_005",
    slug: "abstract-geometry",
    url: getStorageOriginalUrl("abstract-geometry"),
    name: "Abstract Geometry",
  },
];

export class ImageService {
  constructor(private readonly images: ImageMeta[] = IMAGES) {}

  list(): ImageMeta[] {
    return [...this.images];
  }

  get(id: string): ImageMeta | undefined {
    return this.images.find((i) => i.id === id);
  }

  /** Pick a random image, optionally avoiding recently used ids. */
  getRandomImage(excludeIds: string[] = []): ImageMeta {
    const pool =
      this.images.filter((i) => !excludeIds.includes(i.id)) ?? [];
    const choices = pool.length > 0 ? pool : this.images;
    return choices[Math.floor(Math.random() * choices.length)];
  }
}

/** Logical canvas size shared by every source SVG (viewBox 0 0 800 800). */
export const SOURCE_VIEWBOX = 800;

export const imageService = new ImageService();

File: backend/app/services/image_service.py
"""ImageService — configurable image source, metadata only."""
from __future__ import annotations

import random
from dataclasses import dataclass


@dataclass(frozen=True)
class ImageMeta:
    id: str
    url: str
    name: str
    slug: str = ""


IMAGES: list[ImageMeta] = [
    ImageMeta("image_001", "/images/image_001.svg", "Cosmic Fox", "cosmic-fox"),
    ImageMeta("image_002", "/images/image_002.svg", "Futuristic City", "futuristic-city"),
    ImageMeta("image_003", "/images/image_003.svg", "Space Explorer", "space-explorer"),
    ImageMeta("image_004", "/images/image_004.svg", "Neon Cyberpunk", "neon-cyberpunk"),
    ImageMeta("image_005", "/images/image_005.svg", "Abstract Geometry", "abstract-geometry"),
]


class ImageService:
    def __init__(self, images: list[ImageMeta] | None = None):
        self._images = images or IMAGES

    def list(self) -> list[ImageMeta]:
        return list(self._images)

    def get(self, image_id: str) -> ImageMeta | None:
        return next((i for i in self._images if i.id == image_id), None)

    def get_random_image(self, exclude_ids: list[str] | None = None) -> ImageMeta:
        pool = [i for i in self._images if i.id not in (exclude_ids or [])]
        choices = pool or self._images
        return random.choice(choices)


image_service = ImageService()
SOURCE_VIEWBOX = 800
